### PR TITLE
Explorer: Avoid unecessary copies when a value binding is created from a value expression

### PR DESCRIPTION
**Pull request by Pixep**
#### Functional changes

* Avoid unnecessary copies when a value binding is created from a value expression in call parameters
* Ensure the result of the value expression bound is destroyed
* Provide storage to initializing expressions used in call parameters

### CODE REVIEW

File: explorer/interpreter/interpreter.cpp [hunk 1]
```diff
@@ -1718,46 +1715,58 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
         auto iter = call.witnesses().begin();
-        std::advance(iter, act.pos() - 2);
+        std::advance(iter, act.pos() - 1 - num_args);
         return todo_.Spawn(std::make_unique<WitnessAction>(
             cast<Witness>(iter->second), call.source_loc()));
-      } else if (act.pos() == 2 + static_cast<int>(num_witnesses)) {
+      } else if (act.pos() == function_call_pos) {
         //    { { v2 :: v1([]) :: C, E, F} :: S, H}
         // -> { {C',E',F'} :: {C, E, F} :: S, H}
+        // Prepare parameters tuple
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
        // Prepare parameters tuple.
```

File: explorer/interpreter/interpreter.cpp [hunk 2]
```diff
@@ -1718,46 +1715,58 @@ auto Interpreter::StepExp() -> ErrorOr<Success> {
...
+        std::vector<Nonnull<const Value*>> param_values;
+        for (int i = 1; i <= num_args; ++i) {
+          param_values.push_back(act.results()[i]);
+        }
+        const auto* param_tuple = arena_->New<TupleValue>(param_values);
+        // Prepare witnesses
```
Comment (jonmeow) [inline explorer/interpreter/interpreter.cpp]:
```suggestion
        // Prepare witnesses.
```

**Review by jonmeow — APPROVED** [review]
Looks great! Just a couple really minor comment nits.

@zygoloid I'm not sure there's anything useful to review in the trace output here, but letting you know in case I'm missing something.

**Comment by zygoloid** [comment]
> @zygoloid I'm not sure there's anything useful to review in the trace output here, but letting you know in case I'm missing something.

Trace output changes look fine. (Though there's a merge conflict there now.)

**Pull request merged**